GitExtUtils/FileLogging/AsynchronousLogWriter.cs

```diff
@@ -0,0 +1,71 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.IO;
+using System.Text;
+using System.Threading;
+
+namespace GitExtUtils.FileLogging
+{
+    public class AsynchronousLogWriter
+    {
+        private readonly LogFilenameCreator _logFilenameCreator;
+        private readonly ILogFormatter _formatter;
+        private readonly TimeSpan _flushInterval;
+        private readonly CancellationToken _cancellationToken;
+        private readonly ConcurrentQueue<LogEntry> _unsaved = new ConcurrentQueue<LogEntry>();
+
+        public AsynchronousLogWriter(
+            LogFilenameCreator logFilenameCreator,
+            ILogFormatter formatter,
+            TimeSpan flushInterval,
+            CancellationToken cancellationToken)
+        {
+            _logFilenameCreator = logFilenameCreator;
+            _formatter = formatter;
+            _flushInterval = flushInterval;
+            _cancellationToken = cancellationToken;
+        }
+
+        public void Enqueue(LogEntry entry)
+        {
+            _unsaved.Enqueue(entry);
+        }
+
+        public void Run()
+        {
+            if (!Directory.Exists(_logFilenameCreator.LogPath))
+            {
+                Directory.CreateDirectory(_logFilenameCreator.LogPath);
+            }
+
+            while (!_cancellationToken.IsCancellationRequested)
+            {
+                _cancellationToken.WaitHandle.WaitOne(_flushInterval);
+
+                WriteUnsaved();
+            }
+        }
+
+        private void WriteUnsaved()
+        {
+            int max = _unsaved.Count;
+            if (max <= 0)
+            {
+                return;
+            }
+
+            var sb = new StringBuilder(max * 1000);
+            int i = 0;
+            while (i++ < max && _unsaved.TryDequeue(out var entry))
+            {
+                sb.AppendLine(_formatter.Format(entry));
+            }
+
+            using (var file = new FileStream(_logFilenameCreator.Create(), FileMode.Append, FileAccess.Write, FileShare.Read))
```

Comment:
`..._logFilenameCreator.Create()...` looks a bit odd, it suppose to be a full path, no a "name of file".
Perhaps it could be expressed as:

```suggestion
            using (var file = new FileStream(_logFullPathProvider.Get(), FileMode.Append, FileAccess.Write, FileShare.Read))
```
or
```suggestion
            using (var file = new FileStream(_logPathGenerator.Generate(), FileMode.Append, FileAccess.Write, FileShare.Read))
```



In both cases we'll need to rename the underlying API to match.

Reply:
or `GenerateFullPath()` ?